swagger_zipkin/otel_decorator.py

```diff
@@ -0,0 +1,165 @@
+    import pyramid.request.Request   # type: ignore # noqa: F401
+
+T = TypeVar('T', covariant=True)
+P = ParamSpec('P')
+
+tracer = trace.get_tracer("otel_decorator")
+
+
+class OtelResourceDecorator:
+    """A wrapper to the swagger resource.
+
+    :param resource: A resource object. eg. `client.pet`, `client.store`.
+    :type resource: :class:`swaggerpy.client.Resource` or :class:`bravado_core.resource.Resource`
+    """
+
+    def __init__(self, resource: Client, client_identifier: str, smartstack_namespace: str) -> None:
+        self.resource = resource
+        self.client_identifier = client_identifier
+        self.smartstack_namespace = smartstack_namespace
+
+    def __getattr__(self, name: str) -> Resource:
+        return decorate_client(self.resource, self.with_headers, name)
+
+    def with_headers(self, call_name: str, *args: Any, **kwargs: Any) -> Any:
+        kwargs.setdefault('_request_options', {})
+        request_options: dict = kwargs['_request_options']
+        request_options.setdefault('headers', {})
+
+        # what is the right way to get the Request object. can we use contruct_request
+        # https://github.com/Yelp/bravado/blob/master/bravado/client.py#L283C5-L283C22
+        # this would create a bravado dependency.
+        request = get_pyramid_current_request()
+        http_route = getattr(request, "matched_route", "")
+        http_request_method = getattr(request, "method", "")
+
+        parent_span = trace.get_current_span()
+        span_name = f"{http_request_method} {http_route}"
+        with tracer.start_as_current_span(
+            span_name, kind=trace.SpanKind.CLIENT
+        ) as span:
+            self.inject_otel_headers(kwargs, span)
+            self.inject_zipkin_headers(kwargs, span, parent_span)
+
+            # ideally the exception should be scoped for self.with_headers
+            # handle_exception should handle general excetion and the specific exception HTTPError
+            # But this would create a dependency on bravado package. Is this ok?
+            # Assuming resource.operation does throw HTTPError
+            # https://github.com/Yelp/bravado/blob/master/bravado/exception.py
+            with self.handle_exception():
+                span.set_attribute("url.path", getattr(request, "path", ""))
+                span.set_attribute("http.route", http_route)
+                span.set_attribute("http.request.method", http_request_method)
+
+                span.set_attribute("client.namespace", self.client_identifier)
+                span.set_attribute("peer.service", self.smartstack_namespace)
+                span.set_attribute("server.namespace", self.smartstack_namespace)
+                span.set_attribute("http.response.status_code", "200")
+
+                return getattr(self.resource, call_name)(*args, **kwargs)
+
+    @contextmanager
+    def handle_exception(
+        self,
+    ) -> Any:
+        try:
+            yield
+        except Exception as e:
```

Comment:
Would like to catch HTTPError here: https://github.com/Yelp/bravado/blob/master/bravado/exception.py#L56
so we can get the right http status code on failures. Is this the right approach?

Reply:
Done. 
Note: i'm handling the general exception in the contextmanager. But had to move the specific exception HTTPError in the try block. This is because in the handle_exception i did not have access to the current span for setting the error attributes. Doing a trace.get_current_span() always returned the parent span and i tried a few other options which did not work either. So did not feel that pursuit was worth it.